[CL-2909] Execute db:migrate during BE deployment only if there are pending migrations.

adessy opened the pull request.

adessy committed:
[CL-2909] Execute db:migrate only if necessary in CI

Skipping db:migrate during back-end deployment when there are no pending migrations.

adessy committed:
[CL-2909] Rename db_enhancements.rake

kogre reviewed (commented):
If I get it right, in this solution we:

1. Check whether the `public` schema has pending migrations
2. If it has, launch `db:migrate`, which will run the migrations for all tenants

I'm not fully convinced this is the way to go, because if I get it right there is a real risk that a tenant is not in the right database shape, and we wouldn't notice. Also if migrations fail (and not on the `public` schema) we couldn't retry by doing another release.

I think it would better to check the migrations state of all tenants, preferably in one query, and still run `db:migrate` if one or more have pending migrations.

wdyt?

adessy commented:
I have my doubts myself, but I think the simplicity of the solution might outweigh the cons in this case. I'm not huge fan of querying `schema_migrations` directly as it's more an implementation detail than a proper interface. Also, it removes us the ability to retry migrations through CI, but for:
> if I get it right there is a real risk that a tenant is not in the right database shape, and we wouldn't notice.

I don't see how it's any different from today. Will try to find some middle ground.

kogre commented:
It's true that the migrations table is not a public rails interface, but it's so well established that the risk of it breaking is really very small imo. Also, the implementation in the rails source-code is just a few lines, so it's no work to update it it does, or just fall back to db:migrate.

adessy committed:
[CL-2909] Rework db:migrate_if_pending task

The previous solution only verified if all migrations had been executed in the public schema. This new approach ensures that there are no pending migrations across all schemas.

kogre reviewed (approved):
Yay, thanks! :)

adessy committed:
Merge branch 'master' into CL-2913/speed-up-deploy/skip-when-no-pending-migrations

# Conflicts:
#	.circleci/config.yml
